x/concentrated-liquidity/swaps.go

```diff
@@ -1,34 +1,317 @@
+	swapState := SwapState{
+		amountSpecifiedRemaining: tokenAmountInAfterFee,
+		amountCalculated:         sdk.ZeroDec(),
+		sqrtPrice:                curSqrtPrice,
+		tick:                     types.PriceToTick(curSqrtPrice.Power(2)),
+		liquidity:                p.GetLiquidity(),
+	}
+
+	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 6)) && !swapState.sqrtPrice.Equal(sqrtPriceLimit) {
+		sqrtPriceStart := swapState.sqrtPrice
+		nextTick, ok := k.NextInitializedTick(ctx, poolId, swapState.tick.Int64(), zeroForOne)
+		if !ok {
+			return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, sdk.Dec{}, fmt.Errorf("there are no more ticks initialized to fill the swap")
+		}
+
+		nextSqrtPrice, err := types.TickToSqrtPrice(sdk.NewInt(nextTick))
+		if err != nil {
+			return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, sdk.Dec{}, fmt.Errorf("could not convert next tick (%v) to nextSqrtPrice", sdk.NewInt(nextTick))
+		}
+
+		var sqrtPriceTarget sdk.Dec
+		if zeroForOne && nextSqrtPrice.LT(sqrtPriceLimit) || !zeroForOne && nextSqrtPrice.GT(sqrtPriceLimit) {
+			sqrtPriceTarget = sqrtPriceLimit
+		} else {
+			sqrtPriceTarget = nextSqrtPrice
+		}
+
+		sqrtPrice, amountIn, amountOut := types.ComputeSwapStep(
+			swapState.sqrtPrice,
+			sqrtPriceTarget,
+			swapState.liquidity,
+			swapState.amountSpecifiedRemaining,
+			zeroForOne,
+		)
+		swapState.sqrtPrice = sqrtPrice
+
+		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.Sub(amountIn)
+		swapState.amountCalculated = swapState.amountCalculated.Add(amountOut)
+
+		// if we have moved to the next tick,
+		if nextSqrtPrice.Equal(sqrtPrice) {
+			liquidityDelta, err := k.CrossTick(ctx, p.GetId(), nextTick)
+
+			if err != nil {
+				return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, sdk.Dec{}, err
+			}
+			if zeroForOne {
+				liquidityDelta = liquidityDelta.Neg()
+			}
+
+			swapState.liquidity = swapState.liquidity.Add(liquidityDelta)
+
+			if swapState.liquidity.LTE(sdk.ZeroDec()) || swapState.liquidity.IsNil() {
+				return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, sdk.Dec{}, err
+			}
+			if zeroForOne {
+				swapState.tick = sdk.NewInt(nextTick - 1)
+			} else {
+				swapState.tick = sdk.NewInt(nextTick)
+			}
+		} else if !sqrtPriceStart.Equal(sqrtPrice) {
+			swapState.tick = types.PriceToTick(sqrtPrice.Power(2))
+		}
+	}
+
+	// coin amounts require int values
+	// we truncate at last step to retain as much precision as possible
+	amt0 := tokenAmountInAfterFee.Add(swapState.amountSpecifiedRemaining).TruncateInt()
+	amt1 := swapState.amountCalculated.TruncateInt()
+
+	tokenIn = sdk.NewCoin(tokenInMin.Denom, amt0)
+	tokenOut = sdk.NewCoin(tokenOutDenom, amt1)
+
+	return tokenIn, tokenOut, swapState.tick, swapState.liquidity, swapState.sqrtPrice, nil
+}
+
+// TODO: implement this
+func (k Keeper) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, swapFee sdk.Dec, priceLimit sdk.Dec, poolId uint64) (tokenOut sdk.Coin, err error) {
```

Comment:
Ditto: was this just moved with no changes?

Reply:
Yes, no change